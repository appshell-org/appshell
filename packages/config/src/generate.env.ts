import fs from 'fs';
import path from 'path';
import readline from 'readline';

const shouldProcess = (line: string, prefixRegex: RegExp) =>
  !line.match(/^(\s*|#.*)$/) && prefixRegex.test(line);

export default async (env: string, prefix = '') =>
  new Promise<Map<string, string | number | undefined>>((resolve, reject) => {
    const map = new Map<string, string | number | undefined>();
    const dotenvPath = path.resolve(env);

    if (fs.existsSync(dotenvPath)) {
      const prefixRegex = new RegExp(`^${prefix}.*$`);
      const stream = fs.createReadStream(dotenvPath);
      const rl = readline.createInterface({
        input: stream,
      });

      rl.on('line', (line) => {
        if (shouldProcess(line, prefixRegex)) {
          const [NAME, VALUE] = line.split('=');
          let currentValue = process.env[NAME];
          if (!currentValue) {
            currentValue = VALUE;
          }
          map.set(NAME, currentValue);
        }
      });

      rl.on('close', () => {
        resolve(map);
      });
    } else {
      reject(new Error(`${env} not found.`));
    }
  });
