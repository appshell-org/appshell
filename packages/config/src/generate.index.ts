/* eslint-disable no-console */
import { AppshellIndex, AppshellManifest } from './types';
import loadJson from './utils/loadJson';

export default async (registries: string[]): Promise<AppshellIndex> => {
  if (registries.length < 1) {
    console.log(`No registries found. skipping index generation.`);
    return {};
  }

  console.log(`generating appshell index --registry=${registries}`);

  try {
    const manifests = await Promise.all(registries.map(loadJson<AppshellManifest>)).then((items) =>
      items.flat(),
    );
    console.log(
      `Generating index from ${manifests.length} manifest${manifests.length === 1 ? '' : 's'}`,
    );

    return manifests.reduce((entries, manifest) => {
      const res = Object.entries(manifest.remotes).reduce((acc, [key, val]) => {
        acc[key] = val.manifestUrl;

        return acc;
      }, entries);

      return res;
    }, {} as Record<string, string>);

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (err: any) {
    console.error('Error generating appshell index', err.message);
  }

  return {};
};
