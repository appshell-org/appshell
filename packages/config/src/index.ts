/**
 * @appshell/config package API
 */
export { default as configmap } from './configmap';
export { default as deregister } from './deregister';
export { default as generateEnv } from './generate.env';
export { default as generateGlobalConfig } from './generate.global-config';
export { default as generateManifest } from './generate.manifest';
export { default as outdated } from './outdated';
export { default as register } from './register';
export { default as sync } from './sync';
export type {
  AppshellConfig,
  AppshellConfigRemote,
  AppshellGlobalConfig,
  AppshellIndex,
  AppshellManifest,
  AppshellRemote,
  AppshellTemplate,
  Metadata,
  PackageSpec,
  Schema,
  SharedModuleSpec,
} from './types';
export * as utils from './utils';
export * as validators from './validators';
