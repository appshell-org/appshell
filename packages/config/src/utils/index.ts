export { default as copy } from './copy';
export { default as list } from './list';
export { default as load, loadAll } from './load';
export { default as merge } from './merge';
