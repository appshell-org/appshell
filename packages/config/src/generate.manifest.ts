/* eslint-disable no-param-reassign */
import fs from 'fs';
import path from 'path';
import configmap from './configmap';
import { toAppshellManifest } from './mappers/appshell.config';
import { AppshellConfig } from './types';

/**
 * Generates an appshell manifest
 * @param configPath path of the appshell config to compile into an appshell manifest
 * @param webpackConfig path of the webpack config
 * @returns an appshell manifest
 */
export default async <TMetadata extends Record<string, unknown>>(configPath: string) => {
  if (!fs.existsSync(configPath)) {
    // eslint-disable-next-line no-console
    console.log(`No config file found '${configPath}'`);

    return null;
  }

  const file = fs.readFileSync(path.resolve(configPath), 'utf-8');
  const template = JSON.parse(file) as AppshellConfig;
  const configMap = configmap.create(template);

  return toAppshellManifest<TMetadata>(template, configMap);
};
