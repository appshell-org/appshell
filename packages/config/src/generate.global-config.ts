/* eslint-disable no-console */
import { AppshellGlobalConfig } from './types';
import { isValidUrl, merge } from './utils';
import loadJson from './utils/loadJson';
import { AppshellGlobalConfigValidator } from './validators';

export type GenerateGlobalConfigOptions = {
  insecure: boolean;
  apiKey?: string;
};

export default async (
  registries: string[],
  options: GenerateGlobalConfigOptions = { insecure: false, apiKey: undefined },
): Promise<AppshellGlobalConfig> => {
  const defaultGlobalConfig = { index: {}, metadata: {}, overrides: { environment: {} } };

  if (registries.length < 1) {
    console.log(`No registries found. skipping global configuration generation.`);
    return defaultGlobalConfig;
  }

  console.log(
    `generating global appshell configuration --registry=${JSON.stringify(
      registries,
      null,
      2,
    )} --insecure=${options.insecure} --apiKey=${options.apiKey}`,
  );

  try {
    const configs = await Promise.all(
      registries.map((reg) => {
        const registry = isValidUrl(reg) ? `${reg}/appshell.config.json` : reg;
        console.log(`Loading global configuration from ${registry}`);
        return loadJson<AppshellGlobalConfig>(registry, {
          insecure: options.insecure,
          target: /(.config.json)/i,
          apiKey: options.apiKey,
        });
      }),
    ).then((items) => items.flat());
    console.log(
      `Generating global configuration from ${configs.length} source${
        configs.length === 1 ? '' : 's'
      }`,
    );

    return merge(AppshellGlobalConfigValidator, ...configs) as AppshellGlobalConfig;

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (err: any) {
    console.error('Error generating global appshell configuration', err.message);
  }

  return defaultGlobalConfig;
};
