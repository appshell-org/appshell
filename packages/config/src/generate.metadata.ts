/* eslint-disable no-console */
import { Metadata } from './types';
import { merge } from './utils';
import loadJson from './utils/loadJson';
import { appshell_metadata } from './validators';

export default async <T = Record<string, Metadata>>(registries: string[]): Promise<T> => {
  if (registries.length < 1) {
    console.log(`No registries found. skipping metadata generation.`);
    return {} as T;
  }

  try {
    const metadata = await Promise.all(
      registries.map((reg) => loadJson<T>(reg, /(.metadata.json)/i)),
    ).then((items) => items.flat());
    console.log(
      `Generating metadata from ${metadata.length} source${metadata.length === 1 ? '' : 's'}`,
    );

    return merge(appshell_metadata, ...metadata) as T;

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (err: any) {
    console.error('Error generating appshell metadata', err.message);
  }

  return {} as T;
};
