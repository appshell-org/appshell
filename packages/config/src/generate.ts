import fs from 'fs';
import configmap from './configmap';
import { toAppshellManifest } from './mappers/appshell.config';
import { AppshellManifest } from './types';
import { list, merge } from './utils';
import validator from './validators/appshell.manifest';

/**
 * Compiles all configuration into an appshell manifest
 * @param configsDir directory of configs to compile into an appshell manifest
 * @param searchDepth depth to which app configs should be searched in configsDir
 * @returns an appshell manifest
 */
export default <TMetadata extends Record<string, unknown>>(configsDir: string, searchDepth = 1) => {
  const manifests = list(configsDir, searchDepth, /\.json$/i).map<AppshellManifest<TMetadata>>(
    (configPath) => {
      if (!fs.existsSync(configPath)) {
        throw new Error(`Configuration not found at ${configPath}`);
      }

      const config = JSON.parse(fs.readFileSync(configPath).toString());
      const configMap = configmap.create(config);

      return toAppshellManifest<TMetadata>(config, configMap);
    },
  );

  if (!manifests.length) {
    // eslint-disable-next-line no-console
    console.log(`No config files found in '${configsDir}'`);

    return { remotes: {}, modules: {} };
  }

  return merge<AppshellManifest<TMetadata>>(validator, ...manifests);
};
