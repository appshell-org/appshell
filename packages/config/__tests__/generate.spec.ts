import { keys, values } from 'lodash';
import path from 'path';
import generateEnv from '../src/generate.env';
import generateIndex from '../src/generate.index';
import generate from '../src/generate.manifest';
import generateMetadata from '../src/generate.metadata';
import * as utils from '../src/utils';
import manifest from './assets/appshell.manifest.json';

type TestMetadata = {
  route: string;
  displayName: string;
  displayGroup: string;
  order: number;
  icon: string;
};

describe('generate', () => {
  const packageName = 'config';

  describe('runtime env file', () => {
    const env = `packages/${packageName}/__tests__/assets/test.env`;

    it('should generate the runtime environment js file', async () => {
      const environment = await generateEnv(env);

      expect(Object.fromEntries(environment)).toStrictEqual({
        REGISTRY: 'packages/cli/__tests__/assets/appshell_registry',
        ROOT: 'TestModule/Workspace',
        TEST_ENV_FOO: 'foo',
        TEST_ENV_BAR: 'bar',
      });
    });

    it('should capture only prefixed environment vars when prefix is supplied', async () => {
      const environment = await generateEnv(env, 'TEST_');

      expect(Object.fromEntries(environment)).toStrictEqual({
        TEST_ENV_FOO: 'foo',
        TEST_ENV_BAR: 'bar',
      });
    });

    it('should fail gracefully when env file is not found', async () => {
      await expect(generateEnv('does_not_exist.env')).rejects.toThrow();
    });
  });

  describe('manifest', () => {
    const appshellConfig = path.resolve(
      `packages/${packageName}/__tests__/assets/appshell.config.json`,
    );

    process.env.APPS_TEST_URL = 'http://remote-module.com/remoteEntry.js';
    process.env.RUNTIME_ENV = 'development';
    process.env.RUNTIME_ENV_VERSION = '1.0.0';

    it('should generate an appshell manifest from the appshell config', async () => {
      const config = await generate(appshellConfig);

      expect(config).toMatchSnapshot();
    });

    it('should produce null when config is available to process', async () => {
      const config = await generate(undefined as any);

      expect(config).toBeNull();
    });

    it('should contain all remotes', async () => {
      const config = await generate(appshellConfig);
      const expectedRemotes = keys(manifest.remotes);
      const actualRemotes = keys(config?.remotes);

      expect(expectedRemotes).toEqual(actualRemotes);
    });

    it('should apply environment variables to configuration', async () => {
      const config = await generate(appshellConfig);

      const actualUrls = values(config?.remotes).flatMap((remote) => remote.manifestUrl);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(actualUrls.some((url) => url.includes(process.env.APPS_TEST_URL!))).toBeTruthy();
    });

    it('should capture metadata', async () => {
      const config = await generate<TestMetadata>(appshellConfig);

      expect(values(config?.remotes).flatMap((remote) => remote.metadata)).toHaveLength(3);
    });
  });

  describe('index', () => {
    let consoleSpy: jest.SpyInstance;
    beforeEach(() => {
      consoleSpy = jest.spyOn(console, 'error');
    });

    afterEach(() => {
      consoleSpy.mockRestore();
    });

    const singleIndex = [`packages/${packageName}/__tests__/assets/registries/appshell.index.json`];
    const adjunctRegistries = [
      `packages/${packageName}/__tests__/assets/registries/registry_a`,
      `packages/${packageName}/__tests__/assets/registries/registry_b`,
    ];

    it('should merge valid registry indexes', async () => {
      const index = await generateIndex([...adjunctRegistries, ...singleIndex]);

      expect(index).toMatchSnapshot();
    });

    it('should handle a single index file', async () => {
      const index = await generateIndex(singleIndex);

      expect(index).toMatchSnapshot();
    });

    it('should handle a directory', async () => {
      const index = await generateIndex(adjunctRegistries);

      expect(index).toMatchSnapshot();
    });

    it('should log error and return empty if something goes wrong', async () => {
      jest.spyOn(utils, 'merge').mockImplementationOnce(() => {
        throw new Error('Something went wrong');
      });
      const index = await generateIndex(singleIndex);

      expect(consoleSpy).toHaveBeenCalledWith(
        'Error generating appshell index',
        'Something went wrong',
      );
      expect(index).toMatchObject({});
    });
  });

  describe('metadata', () => {
    let consoleSpy: jest.SpyInstance;
    beforeEach(() => {
      consoleSpy = jest.spyOn(console, 'error');
    });

    afterEach(() => {
      consoleSpy.mockRestore();
    });

    const singleIndex = [
      `packages/${packageName}/__tests__/assets/registries/appshell.metadata.json`,
    ];
    const adjunctRegistries = [
      `packages/${packageName}/__tests__/assets/registries/registry_a`,
      `packages/${packageName}/__tests__/assets/registries/registry_b`,
    ];

    it('should merge valid registry indexes', async () => {
      const index = await generateMetadata([...adjunctRegistries, ...singleIndex]);

      expect(index).toMatchSnapshot();
    });

    it('should handle a single index file', async () => {
      const index = await generateMetadata(singleIndex);

      expect(index).toMatchSnapshot();
    });

    it('should handle a directory', async () => {
      const index = await generateMetadata(adjunctRegistries);

      expect(index).toMatchSnapshot();
    });

    it('should log error and return empty if something goes wrong', async () => {
      jest.spyOn(utils, 'merge').mockImplementationOnce(() => {
        throw new Error('Something went wrong');
      });
      const index = await generateMetadata(singleIndex);

      expect(consoleSpy).toHaveBeenCalledWith(
        'Error generating appshell metadata',
        'Something went wrong',
      );
      expect(index).toMatchObject({});
    });
  });
});
