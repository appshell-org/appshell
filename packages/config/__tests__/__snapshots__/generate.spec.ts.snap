// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate should generate an appshell manifest from a collection of app manifests 1`] = `
{
  "modules": {
    "BarModule": {
      "exposes": {
        "./Bar": "./src/Entry2",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "BarModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
    "BizModule": {
      "exposes": {
        "./Biz": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "BizModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
    "FooModule": {
      "exposes": {
        "./Bing": "./src/Entry1",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "FooModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
  },
  "remotes": {
    "BarModule/Bar": {
      "id": "BarModule/Bar",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Bar App",
        "icon": "ViewList",
        "order": 20,
        "route": "/bar",
      },
      "module": "./Bar",
      "scope": "BarModule",
      "shareScope": "@sharedscope",
      "url": "http://remote-module.com/remoteEntry.js",
    },
    "BizModule/Biz": {
      "id": "BizModule/Biz",
      "metadata": {
        "displayGroup": "auxiliary",
        "displayName": "Biz App",
        "icon": "ViewList",
        "order": 30,
        "route": "/biz",
      },
      "module": "./Biz",
      "scope": "BizModule",
      "shareScope": "@sharedscope",
      "url": "http://localhost:4040/remoteEntry.js",
    },
    "FooModule/Bing": {
      "id": "FooModule/Bing",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "module": "./Bing",
      "scope": "FooModule",
      "shareScope": "@sharedscope",
      "url": "http://remote-module.com/remoteEntry.js",
    },
    "FooModule/Foo": {
      "id": "FooModule/Foo",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "module": "./Foo",
      "scope": "FooModule",
      "shareScope": "@sharedscope",
      "url": "http://remote-module.com/remoteEntry.js",
    },
  },
}
`;

exports[`generate should produce an empty manifest when no configs are available to process 1`] = `
{
  "modules": {},
  "remotes": {},
}
`;
